flex/flex.cpp

```diff
@@ -0,0 +1,164 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2021 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "flex/flex.h"
+#include "flex/vk-flex-data.h"
+
+#include <cassert>
+#include <cstdio>
+#include <cstring>
+
+int node::get_child(unsigned char c, const lang *ctx) const noexcept {
+  int l = children_start;
+  int r = children_end;
+  if (r - l <= 4) {
+    for (int i = l; i < r; i++) {
+      if (ctx->children[2 * i] == c) {
+        return ctx->children[2 * i + 1];
+      }
+    }
+  } else {
+    while (r - l > 1) {
+      int x = (r + l) >> 1;
+      if (ctx->children[2 * x] <= c) {
+        l = x;
+      } else {
+        r = x;
+      }
+    }
+    if (ctx->children[2 * l] == c) {
+      return ctx->children[2 * l + 1];
+    }
+  }
+  return -1;
+}
+
+bool lang::has_symbol(char c) const noexcept {
+  return strchr(flexible_symbols, c) != nullptr;
+}
+
+const char *flex(vk::string_view name, vk::string_view case_name, bool is_female, vk::string_view type, int lang_id, char * const dst_buf, char * const err_buf) {
+  if (name.size() > (1 << 10)) {
+    sprintf(err_buf, "Name has length %zu and is too long in function vk_flex", name.size());
+    return name.data();
```

Comment:
nit

roughly speaking it is unsafe to return data() from string_view, because in general it's not a null terminated string, but const char * may assume it. I suggest using vk::string_view as return value and use size() on calling place.

Reply:
fixed